Theory_assignment_two/assignment_2.md

```diff
@@ -0,0 +1,453 @@
+
+Example:
+
+```python
+ # method overloading
+#
+# class Student:
+#     def __init__(self, m1, m2):
+#         self.m1 = m1
+#         self.m2 = m2
+#
+#     def sum(self, a=None, b=None, c=None):
+#         s = 0
+#         if a!=None and b!=None and c!=None:
+#             s = a + b + c
+#         elif a!=None and b!=None:
+#             s = a + b
+#         else:
+#             s = a
+#         return s
+#
+# studentsum_2 = Student(50,12)
+#
+# print(studentsum_2.sum(4,6,5))
+#
+
+```
+
+## 3. What's inheritance in OOP? 
+
+It is the transfer of the characteristics of a class to other classes that are derived from it.
+
+The types of inheritance in python are:
+
+- Single inheritance – when a derived (child) class inherits from a single parent (base) class
+```python
+#
+# class Animal:
+#     def print_details_1(self):
+#         print("This is parent Animal class method")
+#
+#
+# class Horse(Animal):
+#     def print_details_2(self):
+#         print("This is child Horse class method")
+#
+# object = Horse()
+# object.print_details_1()
+# object.print_details_2()
+```
+- Multiple inheritance – this is where a class is derived from more than one base class
+
+```python
+
+#
+# class Mother:
+#     def mother_method(self):
+#         print("This is parent Mother class method")
+#
+# class Father:
+#     def father_method(self):
+#         print("This is parent Father class method")
+#
+# class Child(Mother, Father):
+#     def child_method(self):
+#         print("This is child Child class method")
+#
+# child_a = Child()
+# child_a.mother_method()
+# child_a.father_method()
+# child_a.child_method()
+#
+```
+- Multilevel inheritance- inheritance when a derived (child) class inherits another derived (child) class
+
+```python
+# # multi-level inheritance when a derived (child) class inherits another derived (child) class
+#
+# class Organism:
+#
+#     alive = True
+#
+# class Animal(Organism):
+#
+#     def eat(self):
+#         print("Animal is eating")
+#
+# class Dog(Animal):
+#
+#     def bark(self):
+#         print("This dog is barking")
+#
+#
+# dog = Dog()
+# print(dog.alive)
+# dog.eat()
+# dog.bark()
+```
+
+Hierarchical inheritance - more than one derived (child) class is created from a single parent (base) class
+
+```python
+class Animal:
+    def run(self, name):
+        self.name = name
+        print(f'{self.name} is running')
+
+class Dog(Animal):
+    def jumping(self):
+        print("Dog is jumping")
+
+class Cat(Animal):
+    def sleeping(self):
+        print("Cat is sleeping")
+
+d = Dog()
+d.run("Dog")
+d.jumping()
+c = Cat()
+c.run("Cat")
+c.sleeping()
+```
+
+## 4. If you had to make a program that could vote for the top three funniest people in the office, how would you do that? How would you make it possible to vote on those people? 
+
+Each person in the office should be only to vote for one person each. They shouldn't be able to vote multiple times. 
```

Comment:
7/10 could explain in more detail the attribute of the employee to measure the 'funny' i.e. employee's hobby, photos or jokes. Missing some main considerations or edge cases handling, can employee edit their vote, what if there is more than one employee having the same vote count?